cdf_fabric_replicator/extractor.py

```diff
@@ -65,93 +108,96 @@ def download_files_from_abfss(self, abfss_directory: str) -> None:
+                        external_id=file.name, 
+                        data_set_id=data_set_id, 
+                        source_created_time=created_time, 
+                        source_modified_time=modified_time, 
+                        overwrite=True)
+
+    def write_time_series_to_cdf(self, data_frame:DataFrame ) -> None:
+        external_ids = data_frame["externalId"].unique()
+        for external_id in external_ids:
+            df = data_frame[data_frame["externalId"] == external_id]
+            state_id = f"{self.config.source.raw_time_series_path}-{external_id}-state"
+            df_to_be_written = None
+            if self.state_store.get_state(state_id)[0] is None:
+                df_to_be_written = df
+            else:
+                latest_written_time = self.state_store.get_state(state_id)[0]
+                df_to_be_written = df[df["timestamp"] > latest_written_time]
+
+            if len(df_to_be_written) > 0:                
+                latest_process_time = df_to_be_written['timestamp'].max()
+
+                df_to_be_written.loc[:, "externalId"] = df_to_be_written.apply(
+                    lambda row: self.config.destination.time_series_prefix + row["externalId"], axis=1
                 )
-
-            if self.config.source.abfss_path:
-                dt = DeltaTable(
-                    self.config.source.abfss_path,
-                    storage_options={"bearer_token": token.token, "user_fabric_endpoint": "true"},
+                df_to_be_written = df_to_be_written.pivot(index="timestamp", columns="externalId", values="value")
+                df_to_be_written.index = pd.to_datetime(df_to_be_written.index)
+                self.client.time_series.data.insert_dataframe(df_to_be_written)
+
+                self.set_state(state_id, latest_process_time)
+
+    def write_event_data_to_cdf(self, file_path: str, token: str, state_id: str) -> None:
+        df = self.convert_lakehouse_data_to_df(file_path, token)
+
+        if str(self.state_store.get_state(state_id)[0]) != str(len(df)):
+            events = self.get_events(df)
+
+            self.client.events.upsert(events)
+            self.run_extraction_pipeline(status="success")
+
+            self.set_state(state_id, str(len(df)))
+
+    def set_state(self, state_id: str, value) -> None:
+        self.state_store.set_state(state_id, value)
+        self.state_store.synchronize()
+
+    def get_events(self, df: DataFrame) -> list[EventWrite]:
+        events = []
+        for row in df.iterrows():
+            new_event = EventWrite(
+                external_id=row[1]["externalId"],
+                start_time=row[1]["startTime"],
+                end_time=row[1]["endTime"],
+                type=row[1]["type"],
+                subtype=row[1]["subtype"],
+                metadata=row[1]["metadata"],
+                description=row[1]["description"],
+                asset_ids=self.get_asset_ids(row[1]["assetExternalIds"]),
```

Comment:
Would be nice to make some of these columns constants. ExternalId for example, was used ~15 times in this class.

Reply:
I agree with a traditional sense if we weren't using python but I am not familiar with doing that or if that is particularly conventional for python users. For example if I did that then we would have a long list of class level constants at the top of the class. as opposed to c# where we define constants in a separate categorized class